File: code_samples/section6/problems/problem6_2/problem6_2.cpp
#include <iostream>
#include <vector>
using namespace std;

/*
    This program demonstrates building a MIN-HEAP "in place" from an existing
    array-like container (std::vector<int>).

    Core heap facts (0-based indexing):
        parent(i) = (i - 1) / 2
        left(i)   = 2*i + 1
        right(i)  = 2*i + 2

    Min-heap invariant:
        arr[parent(i)] <= arr[i]  for all i > 0

    Why buildHeap is efficient:
        - Bottom-up heap construction runs in O(n) time.
        - It starts from the last internal node (n/2 - 1) and heapifies down
          each node moving toward the root.
        - Leaves (indices >= n/2) are already valid heaps of size 1.

    Notes:
        - heapify() here is recursive. For very large heaps, an iterative
          version can avoid deep recursion.
        - popMinFromHeapArray() is a testing helper, not part of the core
          "buildHeap" requirement.
*/

static void heapify(vector<int> &arr, int n, int i) {
    /*
        heapify (a.k.a. sift-down)
        --------------------------
        Ensures the subtree rooted at index i satisfies the min-heap property,
        assuming its children subtrees are already heaps.

        Parameters:
            arr : vector storing heap in array form
            n   : number of valid heap elements (heap size)
            i   : index of the node to sift down

        Algorithm:
            1) Identify the smallest among node i and its children.
            2) If one child is smaller than arr[i], swap with the smallest child.
            3) Recurse into the child index we swapped with, since that subtree
               may now violate the heap property.

        Complexity:
            O(log n) worst case (height of heap).
    */
    int smallest = i;
    int left = 2 * i + 1;
    int right = 2 * i + 2;

    // If left child exists and is smaller than current smallest, update smallest.
    if (left < n && arr[left] < arr[smallest]) smallest = left;

    // If right child exists and is smaller than current smallest, update smallest.
    if (right < n && arr[right] < arr[smallest]) smallest = right;

    // If the smallest is not the current node, swap and continue heapifying downward.
    if (smallest != i) {
        swap(arr[i], arr[smallest]);
        heapify(arr, n, smallest);
    }
}

void buildHeap(vector<int> &arr) {
    /*
        buildHeap
        ---------
        Converts an arbitrary vector into a min-heap in place.

        Strategy (bottom-up):
            - Compute n = arr.size()
            - Start from the last internal node: n/2 - 1
            - Heapify down each node going backwards to the root (index 0)

        Why start at n/2 - 1?
            - All indices >= n/2 are leaves (no children), so they are already heaps.
            - Internal nodes may violate the heap property, so we fix them bottom-up.

        Complexity:
            O(n) overall.
    */
    int n = (int)arr.size();

    // Iterate over all internal nodes from last parent down to root.
    for (int i = n / 2 - 1; i >= 0; i--) {
        heapify(arr, n, i);
    }
}

static int popMinFromHeapArray(vector<int> &arr, int &n) {
    /*
        popMinFromHeapArray
        -------------------
        Testing helper: removes and returns the minimum element from an array-based heap.

        Parameters:
            arr : vector storing the heap array
            n   : current heap size (passed by reference and decremented)

        Returns:
            - root (minimum) if n > 0
            - 0 if the heap is empty

        Important caveat:
            Returning 0 for empty is ambiguous if 0 could be a valid element.
            For robust code, prefer:
                - bool pop(int& out),
                - std::optional<int>,
                - or throwing on empty.
    */

    // Simple pop from array-based min-heap (not part of spec, just for testing)
    if (n == 0) return 0;

    // The minimum element is always at index 0 in a min-heap.
    int root = arr[0];

    // Shrink heap size by one (logically removing last element).
    n--;

    if (n > 0) {
        // Move the last element into the root position.
        // This may violate the heap property at the root, so heapify down.
        arr[0] = arr[n];
        heapify(arr, n, 0);
    }

    return root;
}

static void printArray(const string &label, const vector<int> &arr) {
    /*
        printArray
        ----------
        Utility function to print all elements of a vector with a label.
        (When printing a heap array, remember it's "heap order", not sorted order.)
    */
    cout << label;
    for (int x : arr) cout << x << " ";
    cout << "\n";
}

int main() {
    /*
        Demo / test harness

        Steps:
            1) Create an unsorted vector.
            2) Print it.
            3) Call buildHeap() to convert it into a min-heap in place.
            4) Print the resulting heap array representation.
            5) Pop repeatedly from a copy to confirm values come out in ascending order.
    */
    vector<int> arr1 = {9, 4, 7, 1, 0, 3, 2};
    int n1 = (int)arr1.size();

    printArray("Original array: ", arr1);

    // Transform arr1 into a min-heap
    buildHeap(arr1);

    // This print shows heap layout in array form, not sorted order.
    printArray("After buildHeap (min-heap array): ", arr1);

    // Optional correctness test:
    // Repeatedly extracting min should yield a sorted ascending sequence.
    cout << "Extracting elements in sorted order: ";
    int size = n1;

    // Use a temporary copy since extraction mutates the heap structure
    vector<int> heapCopy = arr1;

    while (size > 0) {
        int x = popMinFromHeapArray(heapCopy, size);
        cout << x << " ";
    }
    cout << "\n";

    return 0;
}

File: code_samples/section6/problems/problem6_2/problem6_2.cs
#nullable enable

/*
    This file demonstrates building a MIN-HEAP in-place from an integer array,
    then validating correctness by repeatedly popping the minimum from a COPY
    of the heap.

    Key heap facts (0-based indexing):
        parent(i) = (i - 1) / 2
        left(i)   = 2*i + 1
        right(i)  = 2*i + 2

    Min-heap invariant:
        arr[parent(i)] <= arr[i] for all i > 0

    Structure:
        1) Helper functions for printing, popping min, and heapifying down
           (implemented as top-level local functions).
        2) A simple test harness that:
            - prints original array
            - calls HeapUtils.BuildHeap(arr) to heapify in place
            - prints heap array (heap order, not sorted)
            - extracts mins from a clone to produce sorted output
        3) HeapUtils class that contains:
            - Heapify (recursive sift-down for a single subtree)
            - BuildHeap (bottom-up heap construction, O(n))

    Notes about #nullable enable:
        - This code does not use nullable reference types explicitly, but enabling
          it is fine and can help catch issues in larger programs.
*/

// ===== Helpers =====

// Print helper
void PrintArray(string label, int[] arr) =>
    // Prints: <label><space-separated elements>
    // Example: "Original array: 9 4 7 1 0 3 2"
    Console.WriteLine(label + string.Join(" ", arr));

// Correct popMin for array-based min-heap using prefix size
int PopMin(int[] arr, ref int size) {
    /*
        PopMin
        ------
        Removes and returns the minimum element from an array that is assumed
        to be a min-heap over the prefix [0, size).

        Why "prefix size" matters:
            - The heap may occupy only the first "size" elements of the array.
            - We must NOT use arr.Length as the heap size during extraction,
              because extraction shrinks the heap.

        Parameters:
            arr  : array containing heap elements in indices [0, size)
            size : (by ref) current heap size; decremented after removal

        Returns:
            - int.MinValue if heap is empty (size == 0)
            - otherwise the removed minimum

        Caveat:
            - int.MinValue is used as an "empty" sentinel. This is fine for
              testing, but ambiguous if int.MinValue could be a legitimate value
              stored in the heap. A production API might instead use:
                  * bool TryPop(out int value)
                  * or int? (nullable)
                  * or throw an exception on empty
    */
    if (size == 0) return int.MinValue;

    // In a min-heap, the minimum element is always at index 0.
    int root = arr[0];

    // Reduce heap size by one (logically removing last element).
    size--;

    if (size > 0) {
        // Move the last element in the heap prefix to the root position.
        // This may violate heap ordering at the root, so we sift down.
        arr[0] = arr[size];
        HeapifyDown(arr, size, 0);
    }

    return root;
}

// Local heapify-down using given size (not arr.Length)
void HeapifyDown(int[] arr, int size, int i) {
    /*
        HeapifyDown (iterative sift-down)
        --------------------------------
        Restores the min-heap invariant starting from index i, assuming that
        the subtrees below i are already heaps.

        Parameters:
            arr  : array containing heap data
            size : heap size over prefix [0, size)
            i    : index to sift down from (commonly 0)

        Algorithm:
            - Repeatedly compare node i with its left/right children.
            - Swap with the smallest child if that child is smaller.
            - Continue from the swapped child index.
            - Stop when node i is <= both children, or i has no children.

        Complexity:
            O(log n) worst case (heap height).
    */
    while (true) {
        int smallest = i;
        int left = 2 * i + 1;
        int right = 2 * i + 2;

        // If left child exists within heap prefix and is smaller, mark it.
        if (left < size && arr[left] < arr[smallest]) smallest = left;

        // If right child exists within heap prefix and is smaller, mark it.
        if (right < size && arr[right] < arr[smallest]) smallest = right;

        // If no child is smaller, heap property holds at i.
        if (smallest == i) break;

        // Swap current node with the smaller child.
        (arr[i], arr[smallest]) = (arr[smallest], arr[i]);

        // Continue sifting down from the child position.
        i = smallest;
    }
}

// ===== Your Test =====

/*
    Test input:
        An arbitrary unsorted array.

    We will:
        1) Print it
        2) Heapify it in-place using HeapUtils.BuildHeap
        3) Print the heap array (heap order, not sorted)
        4) Clone the heap and repeatedly PopMin to verify ascending output
*/
int[] arr = [9, 4, 7, 1, 0, 3, 2];

PrintArray("Original array: ", arr);

// BuildHeap transforms the array into a valid min-heap in-place.
HeapUtils.BuildHeap(arr);

// Note: This printed array is in heap layout, not sorted order.
PrintArray("After BuildHeap (min-heap array): ", arr);

// Extract values to test correctness
Console.WriteLine("Extracting elements in sorted order:");

// Copy the heap because PopMin mutates the array contents and shrinks "size".
int[] heapCopy = (int[])arr.Clone();
int size = heapCopy.Length;

// Collect extracted values so we can print them neatly afterwards.
List<int> output = [];

while (size > 0) {
    // PopMin removes the current minimum and reduces "size" by 1.
    int v = PopMin(heapCopy, ref size);
    output.Add(v);
}

// Expected result for the sample input: 0 1 2 3 4 7 9
PrintArray("", output.ToArray());

class HeapUtils {
    /*
        HeapUtils provides:
          - Heapify: a recursive heapify-down for a subtree
          - BuildHeap: bottom-up heap construction in O(n)

        This matches the classic heap construction approach:
          - Start from last internal node (n/2 - 1)
          - Heapify each node going backwards to the root
    */

    private static void Heapify(int[] arr, int n, int i) {
        /*
            Heapify (recursive)
            -------------------
            Ensures the subtree rooted at i satisfies the min-heap property,
            given that its children subtrees already satisfy the property.

            Parameters:
                arr : heap array
                n   : heap size (for build phase this is arr.Length)
                i   : subtree root index

            Steps:
                - pick smallest among i, left child, right child
                - if smallest isn't i, swap and recurse into smallest

            Complexity:
                O(log n) worst case.
        */
        int smallest = i;
        int left = 2 * i + 1;
        int right = 2 * i + 2;

        if (left < n && arr[left] < arr[smallest]) smallest = left;
        if (right < n && arr[right] < arr[smallest]) smallest = right;

        if (smallest != i) {
            // Swap current node with smallest child
            (arr[smallest], arr[i]) = (arr[i], arr[smallest]);

            // Continue heapifying at the child position
            Heapify(arr, n, smallest);
        }
    }

    public static void BuildHeap(int[] arr) {
        /*
            BuildHeap
            ---------
            Converts an arbitrary array into a min-heap in-place.

            Why start at n/2 - 1?
                - Indices >= n/2 are leaves (no children) and already valid heaps.
                - Internal nodes may violate the heap invariant and must be heapified.

            Complexity:
                O(n) overall.
        */
        int n = arr.Length;

        // Heapify from the last internal node down to the root.
        for (int i = n / 2 - 1; i >= 0; i--) {
            Heapify(arr, n, i);
        }
    }
}
